pkg/types/types.go

```diff
@@ -70,7 +70,10 @@ type Address []byte
 
 // UpkeepKey is an identifier of an upkeep at a moment in time, typically an
 // upkeep at a block number
-type UpkeepKey []byte
+type UpkeepKey interface {
+	BlockKeyAndUpkeepID() (BlockKey, UpkeepIdentifier, error)
+	fmt.Stringer
```

Comment:
This interface should include json marshaler and unmarshaler interfaces

Reply:
I don't think it should? We should never directly call the marshal or unmarshal functions on an UpkeepKey implementation, and for any other implementations that we introduce, it should be implementation-specific whether or not we need to define custom behavior for marshalling and unmarshalling?

Reply:
I think the thought process I was going down was to enforce custom marshaling/unmarshaling on future implementations. Adding the functions to this interface says to future devs "we use JSON encoding and you need to implement custom functions for it".

Reply:
Right, but we shouldn't need to *enforce* custom implementations for marshalling or unmarshalling? 

Like, if I add a new UpkeepKey implementation, and I don't like the default marshalling/unmarshalling functionality for that type, then I can specify my own Marshal and Unmarshal functions as desired, which is fine. 

But if we add the json marshal and unmarshal interfaces here, then we're saying that every UpkeepKey type *must* implement their own custom functionality for both operations, even in cases where the "default" marshalling and unmarshalling is the desired behavior. 

So, potentially we'd be forcing future developers to explicitly re-write the default marshalling and unmarshalling behavior for certain types? 

That's what I had in an earlier version of this PR, when I added the json.Unmarshaler interface to this interface, and my unmarshal function was just performing the default unmarshal behavior, which was just extra unnecessary code, (which has since been removed).

Reply:
Right, we would be forcing future developers to explicitly re-write default marshaling. The intention I'm after is actually validation. I would like to enforce the idea that the incoming upkeep keys need to be validated. Maybe a validation function instead? I was thinking about unmarshaling as the validation step, but it can be done another way.